contracts/introspection/ERC165Query.sol

```diff
@@ -0,0 +1,83 @@
+pragma solidity ^0.4.20;
+
+
+/**
+ * @title ERC165Query
+ * @dev https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md
+ */
+library ERC165Query {
+  bytes4 constant INVALID_ID = 0xffffffff;
+  bytes4 constant ERC165_ID = 0x01ffc9a7;
+
+  /**
+   * @notice Query if a contract implements an interface
+   * @param _contract The address of the contract to query for support of an interface
+   * @param _interfaceId The interface identifier, as specified in ERC-165
+   * @return true if the contract at _contract indicates support of the interface with
+   * identifier _interfaceId, false otherwise
+   * @dev Interface identification is specified in ERC-165. This function
+   * uses less than 30,000 gas.
+   */
+  function doesContractImplementInterface (
+    address _contract, 
+    bytes4 _interfaceId
+  ) 
+    external 
+    view
+    returns (bool) 
+  {
+    uint256 success;
+    uint256 result;
```

Comment:
Must these be `uint256`? Can't we use `bool`?

Reply:
So this code came from the spec: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md

I take it that the EIP authors used uint256 here as thats what the inline assembly returns by default when you mload.

We can change the noThrowCall code to return bools, but I would prefer the library code here to stay as close to the spec as possible.

Reply:
AFAIK, that code is not technically part of the specification, but a sample contract to test it. And [according to the Solidity docs](https://solidity.readthedocs.io/en/v0.4.24/types.html#members-of-addresses), `staticcall` returns a `bool`, as does `supportsInterface` [according to the 165 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md#how-a-contract-will-publish-the-interfaces-it-implements), so it's not like we're doing anything weird.

Encoding `success` and `result` in `uint256` is confusing IMO, since those are actually boolean values (as evidenced by the fact that they are only compared to 0 and 1). I'd make them `bool` from the get go to reduce noise and make our intent 100% clear. 

Similarly, I'd add comments on the 3 `noThrowCall` calls, indicating what is being done on each one (check that the method exists, check that it returns false for invalid ids, and only then check the actual method that is being queried). I may even go as far as splitting that into two methods, one checking ERC165 support (the first two calls), and one only checking for the requested method (the last one), so that, when checking multiple methods, the first two calls are not needlessly duplicated (saving up to 60k gas per check).

That way, we'll end up with a cleaner query library :)